### PR TITLE
Add a max brightness field for each dimmer in config file #149

### PR DESCRIPTION
**Pull request by AlexPinhasov**
Hi milo I am not sure how can I test it and debug it, I am using VSC, if you could share with me this knowledge I will test and finish this feature fix.

like we discussed I added a new field in config file, first I touch type script language so I hope I added it in the right places

### CODE REVIEW

**Comment by milo526** [comment]
To test this:

Please make sure you have a local installation of homebridge.
In the directory in which you cloned this git repo, run `npm link`
Run `yarn watch` or `npm run watch` depending on which packaging system you use.

Thank you for your work on this! It is much appreciated.

File: config.schema.json [hunk 1]
```diff
@@ -78,6 +78,12 @@
               "description": "Can be found in your platform app (device info -> virtual ID) and is exposed during plugin boot (use the Tuya ID)",
               "required": true
             },
+            "max_brightness": {
+              "title": "The Device max brigthness",
+              "type": "number",
+              "default": "255",
+              "description": "Each dimmer light has different max value, find it and set it here"
```
Comment (milo526) [inline config.schema.json:85]: Please document this in the readme file, specifically how people should find this.

Probably by using the Tuya app to set the value to max and using the API to check the current value?

File: package.json [hunk 2]
```diff
@@ -56,7 +56,7 @@
     "mocha": "^8.2.1",
     "nodemon": "^2.0.6",
     "np": "^7.0.0",
-    "prettier": "^2.2.1",
+    "prettier": "2.2.1",
```
Comment (milo526) [inline package.json:59]: Please revert this change.

**Review by milo526 — CHANGES_REQUESTED** [review]

**Comment by milo526** [comment]
Please also make sure to implement the `validateConfigOverwrites` method on the `DimmerAccessory` and `LightAccessory` classes to validate the provided values.
You can look into the `ClimateAccessory` for how to do this.

File: config.schema.json (hunk 1 above)
Comment (AlexPinhasov) [reply config.schema.json:85]: @milo526 Let me know if the description is good enough

File: package.json (hunk 2 above)
Comment (AlexPinhasov) [reply package.json]: Done

**Comment by AlexPinhasov** [comment]
For some reason running prettier is not enough and the "Build and Lint" fail, any idea what am I missing ?

File: src/accessories/characteristics/brightness.ts [hunk 3]
```diff
@@ -62,8 +62,10 @@ export class BrightnessCharacteristic extends TuyaWebCharacteristic {
       data?.color?.brightness !== undefined
     ) {
       stateValue = Number(data.color.brightness);
-    } else if (data?.brightness) {
-      stateValue = Math.round((Number(data.brightness) / 255) * 100);
+    } else if (data?.brightness && data?.max_brightness) {
+      stateValue = Math.round(
+        (Number(data.brightness) / data.max_brightness) * 100
+      );
```
Comment (milo526) [inline src/accessories/characteristics/brightness.ts:68]: `data?.max_brightness` will not be defined in case the user does not enter device defaults information.
Please use the old value (255) as the default value In this case.

```suggestion
    } else if (data?.brightness) {
      const max_brightness = data?.max_brightness ?? 255
      stateValue = Math.round(
        (Number(data.brightness) / max_brightness) * 100
      );
```

This will use 255 as fallback incase the `max_brightness` is not set.

File: README.md [hunk 4]
```diff
@@ -106,6 +106,9 @@ The `defaults` has these properties:
 
 - `id` The name or id for the device that is registered in the Android/iOS App. When matching on ID please provide the `Tuya ID` as shown during Homebridge boot.
 - `device_type` The `device_type` to be overruled. This can be useful for dimmers that are reported as `light` by the Tuya API and don't support hue and saturation or for outlets that are reported as `switch`.
+- `max_brightness` The maximum value of your dimmer light brightness (It seems that not all dimmers has a the same value), you can get this value by setting your dimmer to 100% in Tuya\Smartlife app, then go into your homebridge main log screen and open the homekit app, once the accessories load you can view the output and see the current value of each light, this output value is your accessories maximum brightness.
```
Comment (milo526) [inline README.md:109]:
```suggestion
- `max_brightness` The maximum value of your dimmer light brightness. To get this, turn your dimmer/light to 100% brightness via the Tuya App. Next open your HomeKit app. Lastly view your HomeBridge logs, once the accessories load you can view the output and see the current value of each light. Something along the lines of `Characteristic.Brightness - [GET] 255` should be shown - in which case you want to set the numerical value of that log entry to the `max_brightness` value.
```

File: README.md [hunk 5]
```diff
@@ -106,6 +106,9 @@ The `defaults` has these properties:
...
+
+[14/02/2021, 19:26:34] [TuyaWebPlatform] [office light] Characteristic.Brightness - [GET] 255
```
Comment (milo526) [inline README.md:111]: These changes are not necessary when combined with the suggested change above.

File: config.schema.json [hunk 6]
```diff
@@ -78,6 +78,12 @@
               "description": "Can be found in your platform app (device info -> virtual ID) and is exposed during plugin boot (use the Tuya ID)",
               "required": true
             },
+            "max_brightness": {
+              "title": "The Device max brigthness",
```
Comment (milo526) [inline config.schema.json:82]:
```suggestion
              "title": "Device max brightness",
```

File: src/api/response.ts [hunk 7]
```diff
@@ -4,6 +4,7 @@ import { TuyaDeviceDefaults } from "../config";
 export type ExtendedBoolean = boolean | "true" | "false" | "True" | "False";
 
 type TuyaProperties = Partial<{
+  max_brightness: number;
```
Comment (milo526) [inline src/api/response.ts:7]: Please move this to the `CustomProperties` below. 
The `TuyaProperties` is used for values that might be provided by the Tuya API which this property isn't.

**Review by milo526 — CHANGES_REQUESTED** [review]

**Comment by milo526** [comment]
This is a great start! Please see some of my feedback, some of which I didn't catch the first time - sorry for that.

File: config.schema.json (hunk 1 above)
Comment (milo526) [reply config.schema.json:85]: See the provided feedback in the relevant file.

File: src/accessories/characteristics/brightness.ts (hunk 3 above)
Comment (AlexPinhasov) [reply src/accessories/characteristics/brightness.ts]: Actually I set the default value for max brightness to be 255 in the config file, then I read that value weather its provided or not, good enough? (check out latests commit)

File: README.md (hunk 4 above)
Comment (AlexPinhasov) [reply README.md:109]: Done

File: README.md (hunk 5 above)
Comment (AlexPinhasov) [reply README.md:111]: Removed

File: config.schema.json (hunk 6 above)
Comment (AlexPinhasov) [reply config.schema.json:82]: Done

File: src/api/response.ts (hunk 7 above)
Comment (AlexPinhasov) [reply src/api/response.ts:7]: Done

**Comment by AlexPinhasov** [comment]
@milo526  I noticed there is another bug, at some point we stop receiving the real value from the accessory numeric scale (0...255) for example, and moving to (0...100) both for update value in HomeKit and reading values from Tuya app.
Once I close the device and wait a few minutes -> foregrounding the HomeKit app we go back to (0...255) weird.
So while this PR fixes the max brightness thats not the only thing.

File: src/accessories/characteristics/brightness.ts (hunk 3 above)
Comment (milo526) [reply src/accessories/characteristics/brightness.ts]: I believe you only set the "default" value in the ui config. This has nothing to do with the actual config file. Just when users use config-ui-x. Furthermore they can still remove the default value, making it undefined. Or not add the device config at all. Therefor I believe the proposed changes are still necessary.

**Comment by milo526** [comment]
> @milo526 I noticed there is another bug, at some point we stop receiving the real value from the accessory numeric scale (0...255) for example, and moving to (0...100) both for update value in HomeKit and reading values from Tuya app.
> Once I close the device and wait a few minutes -> foregrounding the HomeKit app we go back to (0...255) weird.
> So while this PR fixes the max brightness thats not the only thing.

That is due to the caching we provide here: https://github.com/milo526/homebridge-tuya-web/blob/master/src/accessories/characteristics/brightness.ts#L44-L47

Whenever we set the value in HomeKit (through the slider or through fetching data from the API) we cache the last known value. This value is always on the scale from 0 to 100. The characteristic should probably be used to convert the 0...100 scale to 0 to `max_brightness`

File: src/accessories/characteristics/brightness.ts (hunk 3 above)
Comment (AlexPinhasov) [reply src/accessories/characteristics/brightness.ts]: Done

**Comment by AlexPinhasov** [comment]
Can you point me to the correct file?
also
@milo526 what are those magic numbers ?
const value = ((homekitValue as number) / 10) * 9 + 10;

**Comment by milo526** [comment]
> @milo526 what are those magic numbers ?
> const value = ((homekitValue as number) / 10) * 9 + 10;

So this maps the homeKitValue (range 0 - 100) to a range 10 - 100 since Tuya brightness value < 10 will just turn the bulb off in my experience.

**Comment by AlexPinhasov** [comment]
So this should be 
((homekitValue as number) / 100) * max_brightness ?

File: src/accessories/characteristics/brightness.ts (hunk 3 above)
Comment (AlexPinhasov) [reply src/accessories/characteristics/brightness.ts]: data?.max_brightness is null for me, before the change I read the value from the config file, where is the connection between data.max_brightness and config.max_brightness should be set?

**Comment by AlexPinhasov** [comment]
Hi @milo526 why did we stop?

File: src/accessories/DimmerAccessory.ts [hunk 8]
```diff
@@ -42,4 +43,19 @@ export class DimmerAccessory extends BaseAccessory {
 
     return super.deviceSupportedCharacteristics;
   }
+
+  validateConfigOverwrites(config: TuyaDeviceDefaults): string[] {
+    const errors = super.validateConfigOverwrites(config);
+    if (config?.max_brightness) {
+      const maxBrigthness = config.max_brightness;
```
Comment (bamodio) [inline src/accessories/DimmerAccessory.ts:50]: This looks odd to me. We are getting config.max_brightness to only reassign it back to its original value, if it was valid? What's the purpose of that? Same in the other file?

**Comment by bamodio** [comment]
> > @milo526 I noticed there is another bug, at some point we stop receiving the real value from the accessory numeric scale (0...255) for example, and moving to (0...100) both for update value in HomeKit and reading values from Tuya app.
> > Once I close the device and wait a few minutes -> foregrounding the HomeKit app we go back to (0...255) weird.
> > So while this PR fixes the max brightness thats not the only thing.
> 
> That is due to the caching we provide here: https://github.com/milo526/homebridge-tuya-web/blob/master/src/accessories/characteristics/brightness.ts#L44-L47
> 
> Whenever we set the value in HomeKit (through the slider or through fetching data from the API) we cache the last known value. This value is always on the scale from 0 to 100. The characteristic should probably be used to convert the 0...100 scale to 0 to `max_brightness`

I noticed this second bug as well. We have a mismatch in the read/written cached values as they are not consistent in scale. I'm not immediately sure the right spot to fix this though as setRemoteValue() and updateValue() are abstracted from the cache. So somewhere, somehow, we'll need to make this contract more consistent?

**Comment by milo526** [comment]
> Hi @milo526 why did we stop?

Because I left for that evening and then waited for you to implement the provided feedback 🙈
Anything I can still help you with? Do you still want to continue working on this?

**Comment by AlexPinhasov** [comment]
@milo526 Yes why not?, this is a very annoying issue
I need 2 things from you
1) 
_So this maps the homeKitValue (range 0 - 100) to a range 10 - 100 since Tuya brightness value < 10 will just turn the bulb off in my experience._
((homekitValue as number) / 100) * max_brightness ?

2)
data?.max_brightness is null for me, before the change I read the value from the config file, where is the connection between data.max_brightness and config.max_brightness should be set?

File: src/accessories/DimmerAccessory.ts (hunk 8 above)
Comment (milo526) [reply src/accessories/DimmerAccessory.ts:50]: We are missing a cast to number here.
Most config variables are or can be strings, so we need to use
`const maxBrightness = Number(config.max_brightness)`

That also explains why we write it back.

File: src/accessories/LightAccessory.ts [hunk 9]
```diff
@@ -59,4 +60,19 @@ export class LightAccessory extends ColorAccessory {
 
     return super.deviceSupportedCharacteristics;
   }
+
+  validateConfigOverwrites(config: TuyaDeviceDefaults): string[] {
+    const errors = super.validateConfigOverwrites(config);
+    if (config?.max_brightness) {
+      const maxBrigthness = config.max_brightness;
```
Comment (milo526) [inline src/accessories/LightAccessory.ts:67]: Also cast to number (and there is a small typo should be `maxBrightness`)

File: src/accessories/characteristics/brightness.ts [hunk 10]
```diff
@@ -63,7 +63,8 @@ export class BrightnessCharacteristic extends TuyaWebCharacteristic {
     ) {
       stateValue = Number(data.color.brightness);
     } else if (data?.brightness) {
-      stateValue = Math.round((Number(data.brightness) / 255) * 100);
+      const max_brightness = data?.max_brightness ?? 255;
```
Comment (milo526) [inline src/accessories/characteristics/brightness.ts:66]: `max_brightness` should be defined as a getter property on the accessory. It is not part of the data.

Please take a look at the `currentTemperatureFactor` method in the `TemperatureSensorAccessory`.

**Review by milo526 — CHANGES_REQUESTED** [review]

**Comment by milo526** [comment]
> @milo526 Yes why not?, this is a very annoying issue
> I need 2 things from you
> 1)
> _So this maps the homeKitValue (range 0 - 100) to a range 10 - 100 since Tuya brightness value < 10 will just turn the bulb off in my experience._
> ((homekitValue as number) / 100) * max_brightness ?


No, please see the what @bamodio said in https://github.com/milo526/homebridge-tuya-web/issues/149 
We need to ensure that the light is not off when you set it to a low percentage (but still > 0).

Your suggestion would still have this problem in place

> data?.max_brightness is null for me, before the change I read the value from the config file, where is the connection between data.max_brightness and config.max_brightness should be set?

See my comment on the MR, you should get this value from the accessory device config, not from the data.

**Comment by AlexPinhasov** [comment]
So we can assume we should support 2 types of ranges
TUYA_BRIGHTNESS_RANGE0 = (1, 255)
TUYA_BRIGHTNESS_RANGE1 = (10, 1000)
meaning we can add a tick box for each one, default should be 1-255 for all of them
thats what you to do?

**Comment by milo526** [comment]
> So we can assume we should support 2 types of ranges
> TUYA_BRIGHTNESS_RANGE0 = (1, 255)
> TUYA_BRIGHTNESS_RANGE1 = (10, 1000)
> meaning we can add a tick box for each one, default should be 1-255 for all of them
> thats what you to do?

I don't necessarily think so since I seem to vaguely remember that there are also (1,100) ranges.
Please make them user-configurable (both min and max), Make 1 - 255 the default, Document the possible other ranges in at least the readme and possible the config-ui-x

**Comment by AlexPinhasov** [comment]
Ok so final solution:
Add both min + max to settings and use that for the accessory
default should be 1-255
add description for both on readme

thats it?

**Comment by milo526** [comment]
> Ok so final solution:
> Add both min + max to settings and use that for the accessory
> default should be 1-255
> add description for both on readme
> 
> thats it?

Sounds good!
Just make sure that the settings can be correctly configured through the config-ui-x plugin as well as through changing the config.json file

**Comment by AlexPinhasov** [comment]
Finished with the minimum value.
there is something wrong with your cache system, I now for a fact my light is 1-255 every is correct, but the cache uses different values all the time.

**Comment by AlexPinhasov** [comment]
Upon working on the minimum value I came a cross an odd behaiover:
1) the range I set for the device was 10-255
2) the range I get for the device is 25-255

I printed the 1% and 100% of the device and got those results, it gets weird

if I set the value in Homekit for example 62% and I take the minimum as 10 then Tuya app reads it correctly 62% perfect
If I set the value on Tuya app for example 1% the value I get is 25 which suggests this is the minimum of the device

even weirder is that checking in the automation rules in the Tuya app the range is 25-1000 
What is going on

**Comment by AlexPinhasov** [comment]
@bamodio  @milo526 Great news !
I think I found the solution, I tested is on my own devices some has 10-255 range and some 10-1000 both seems to work

only 1 issue remains maybe you can assist, the decimal number sometimes can differ in some cases resulting in accuracy of 1-/+ , I noticed that values above 40 % are correct exactly where less then that we might see a 1% different

Still with all of this its already a big leap in solving this once and for all, give it a test and lets roll out?

File: src/accessories/LightAccessory.ts [hunk 11]
```diff
@@ -11,6 +11,7 @@ import {
 } from "./characteristics";
 import { ColorAccessory } from "./ColorAccessory";
 import { TuyaDevice } from "../api/response";
+import { TuyaDeviceDefaults } from "../config";
```
Comment (milo526) [inline src/accessories/LightAccessory.ts:14]: There are some warnings here, could you please take a look at those.

File: src/accessories/characteristics/brightness.ts [hunk 12]
```diff
@@ -41,7 +42,13 @@ export class BrightnessCharacteristic extends TuyaWebCharacteristic {
     callback: CharacteristicSetCallback
   ): void {
     // Set device state in Tuya Web API
-    const value = ((homekitValue as number) / 10) * 9 + 10;
+    const dimmerPercentage = (homekitValue as number) / 100;
+    const correctionFactor = 10;
+    const value = Math.round(
+      (100 - correctionFactor) * dimmerPercentage +
+        correctionFactor +
+        Number.EPSILON
```
Comment (milo526) [inline src/accessories/characteristics/brightness.ts:50]: Why did you opt to add the `Number.EPSILON` here (and a few lines later)?

**Review by milo526 — CHANGES_REQUESTED** [review]

File: src/accessories/characteristics/brightness.ts (hunk 12 above)
Comment (AlexPinhasov) [reply src/accessories/characteristics/brightness.ts:50]: The decimal number if very large, I wanted to be sure as much as possible the round will be correct

**Comment by milo526** [comment]
Thank you for your work once more.
I'll have some time to check this later this week.
Maybe @bamodio is able to test this in the meantime?

**Comment by AlexPinhasov** [comment]
@milo526 So any news?

**Comment by bamodio** [comment]
I tested this a bit... There's still errors, and I believe they are related to the caching of values, and the mismatch of ranges...

Here's a description and log sample...
1) My config is set to:
            "defaults": [
                {
                    "id": "Gym",
                    "device_type": "light",
                    "max_brightness": 1000,
                    "min_brightness": 10
                }
            ],

2) In Homebridge UI > Accessories, I choose the Gym light and set the brightness to 82%.
3) The config works well and I see this line emitted, so far so good!
**[3/18/2021, 2:22:49 AM] [Tuya] [Gym] Characteristic.Brightness - [SET] 820**

4) Immediately after that though, there's a set of lines that I think are refreshing the value form the cache...
[3/18/2021, 2:22:49 AM] [Tuya] [Gym] - Requesting device state
[3/18/2021, 2:22:49 AM] [Tuya] [Gym] - Creating new debounced promise
[3/18/2021, 2:22:49 AM] [Tuya] [Gym] - Triggering debouncedDeviceStateRequest
[3/18/2021, 2:22:50 AM] [Tuya] [Gym] - Unsetting debouncedDeviceStateRequestPromise
[3/18/2021, 2:22:50 AM] [Tuya] [Gym] - Renewing cache due to RateLimitError
**[3/18/2021, 2:22:50 AM] [Tuya] [Gym] Characteristic.Brightness - [GET] 82**

5) You see that the cache is returning the "Homekit value" (range 1-100) and not the Value I had set above: 820. The expectation here would be that the cache stores this value.

6) I think there's a problem or confusion with the value being cached... updateValue() gets called multiple time with the value 82, and not 820. 

Is this coming from the cache somehow? Is there a disparity of using/caching homekitValue vs the remote value?

This is not working yet, sorry guys.

@milo526 ?

**Comment by AlexPinhasov** [comment]
@bamodio You are correct I have stated before that @milo526 has a problem with caching.
But if you open the Tuya/Smartlife app and change the value then HomeKit will be correct once its not cached, and if you change it via HomeKit the Tuya app is also correct, as far as I am concerned the fix is valid, @milo526 Can you take a look on the caching mechanism ?

**Comment by milo526** [comment]
Please show me where the caching mechanism is wrong in the current plugin.

**Comment by AlexPinhasov** [comment]
@milo526 The scenario is:
1) In home kit change the brightness to 67% then go to Tuya/smartlfe app and see 67% so far so good, when you come back to HomeKit instead of showing 67% you see the a cached number which is incorrect, if you wait long enough and reenter the HomeKit app to fetch the value instead of using the cache it goes back to 67% as it should

**Comment by milo526** [comment]
Hey, I made some time to look at this today but there seems to be some merge conflicts.
I am however not allowed to resolve them for you.

Could you please enable the [Allow edits from maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) option.

**Comment by AlexPinhasov** [comment]
@milo526 Its enabled, didn't touch it

**Comment by milo526** [comment]
Sadly GitHub will not allow me to push any changes at this time.

**Comment by AlexPinhasov** [comment]
@milo526 I have sent you an invite to be a collaborator in my master branch, accept it and try again

**Comment by AlexPinhasov** [comment]
@milo526 You gave up?

**Comment by milo526** [comment]
👍 Yeah that is exactly what I did, I did not spend any time on this at all, I have all the time in the world to work on this and have no other obligations.

I'll fix this myself when I have the time.
Thanks for trying.

**Comment by AlexPinhasov** [comment]
@milo526 Lol, did not expect that.
I guess this was fake -> "Thank you for your work on this! It is much appreciated."
I spent days to solve this issue which obviously effects many users. even with the cache issue this PR would have fixed long brightness issues you have.
Thats the thank you I get for spending my time on your repository.

**Comment by milo526** [comment]
Please feel free to release your own branch as a separate plugin!
